noisy-simulator/src/density_matrix_simulator.rs

```diff
@@ -0,0 +1,346 @@
+            .as_mut()?
+            .apply_operation_matrix(operation.matrix(), qubits)?;
+        if let Err(err) = self.state.as_mut()?.renormalize() {
+            handle_error!(self, err);
+        }
+        Ok(())
+    }
+
+    /// Apply non selective evolution to the given qubit ids.
+    fn apply_instrument(&mut self, instrument: &Instrument, qubits: &[usize]) -> Result<(), Error> {
+        self.check_out_of_bounds_qubits(qubits)?;
+
+        self.state
+            .as_mut()?
+            .apply_operation_matrix(instrument.non_selective_operation_matrix(), qubits)?;
+        if let Err(err) = self.state.as_mut()?.renormalize() {
+            handle_error!(self, err);
+        }
+        Ok(())
+    }
+
+    /// Performs selective evolution under the given instrument.
+    /// Returns the index of the observed outcome.
+    ///
+    /// Use this method to perform measurements on the quantum system.
+    fn sample_instrument(
+        &mut self,
+        instrument: &Instrument,
+        qubits: &[usize],
+    ) -> Result<usize, Error> {
+        let sample = self.rng.gen();
+        self.sample_instrument_with_distribution(instrument, qubits, sample)
+    }
+
+    /// Performs selective evolution under the given instrument.
+    /// Returns the index of the observed outcome.
+    fn sample_instrument_with_distribution(
+        &mut self,
+        instrument: &Instrument,
+        qubits: &[usize],
+        random_sample: f64,
+    ) -> Result<usize, Error> {
+        self.check_out_of_bounds_qubits(qubits)?;
+
+        let mut tmp_state = self.state.clone()?;
+        apply_kernel(
+            &mut tmp_state.data,
+            instrument.total_effect_transposed(),
+            qubits,
+        )?;
+        let total_effect_trace = tmp_state.trace();
+        if total_effect_trace < TOLERANCE {
+            let err = Error::ProbabilityZeroEvent;
+            handle_error!(self, err);
+        }
+        let mut last_non_zero_trace_outcome: usize = 0;
+        let mut last_non_zero_trace: f64 = 0.0;
+        let mut summed_probability: f64 = 0.0;
+
+        for outcome in 0..instrument.num_operations() {
+            if summed_probability > random_sample {
+                break;
+            }
+            tmp_state = self.state.clone()?;
+            apply_kernel(
+                &mut tmp_state.data,
+                instrument.operation(outcome).effect_matrix_transpose(),
+                qubits,
+            )?;
+            let outcome_trace = tmp_state.trace();
+            summed_probability += outcome_trace / total_effect_trace;
+            if outcome_trace >= TOLERANCE {
+                last_non_zero_trace_outcome = outcome;
+                last_non_zero_trace = outcome_trace;
+            }
+        }
+
+        if summed_probability + TOLERANCE <= random_sample || last_non_zero_trace < TOLERANCE {
+            let err = Error::FailedToSampleInstrumentOutcome;
+            handle_error!(self, err);
+        }
```

Comment:
In many of these cases I see handle_error! used where I would normally expect to see map_err. Was this an intentional design choice?

Reply:
Yes, this is an intentional design choice. Because depending on the kind of error the simulator state becomes invalid, and we need to mutate `self.state` to be that last error, then `return err` from the function.